persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,536 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections #-}
+
+module Database.Persist.Quasi.Internal.ModelParser
+    ( SourceLoc (..)
+    , Token (..)
+    , tokenContent
+    , anyToken
+    , ParsedEntityDef
+    , parsedEntityDefComments
+    , parsedEntityDefEntityName
+    , parsedEntityDefIsSum
+    , parsedEntityDefEntityAttributes
+    , parsedEntityDefFieldAttributes
+    , parsedEntityDefFieldComments
+    , parsedEntityDefExtras
+    , parsedEntityDefSpan
+    , parseSource
+    , memberBlockAttrs
+    ) where
+
+import Prelude hiding (lines, unlines)
+import Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NEL
+import qualified Data.Map as M
+import Data.Maybe
+import Data.Text (Text, pack, unlines, unpack)
+import Data.Void
+import Database.Persist.Types
+import Database.Persist.Types.Span
+import Language.Haskell.TH.Syntax (Lift)
+import Replace.Megaparsec (sepCap)
+import Text.Megaparsec hiding (Token)
+import Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer as L
+
+type Parser = Parsec Void String
+
+-- | Source location: file and line/col information. This is half of a 'Span'.
+data SourceLoc = SourceLoc
+    { locFile :: Text
+    , locStartLine :: Int
+    , locStartCol :: Int
+    }
+    deriving (Show, Lift)
+
+data Token
+    = DocComment Text
+    | Comment Text
+    | Quotation Text
+    | Equality Text Text
+    | Parenthetical Text
+    | BlockKey Text
+    | PText Text
+    deriving (Eq, Ord, Show)
+
+tokenContent :: Token -> Text
+tokenContent = \case
+    Comment s -> s
+    DocComment s -> s
+    Quotation s -> s
+    Equality l r -> mconcat [l, "=", r]
+    Parenthetical s -> s
+    PText s -> s
+    BlockKey s -> s
+
+spaceConsumer :: Parser ()
+spaceConsumer =
+    L.space
+        hspace1
+        (L.skipLineComment "--")
+        empty
+
+spaceConsumerN :: Parser ()
+spaceConsumerN =
+    L.space
+        space1
+        (L.skipLineComment "--")
+        empty
+
+escapedParen :: Parser Char
+escapedParen =
+    char '\\'
+        *> choice
+            [ char '('
+            , char ')'
+            ]
+
+contentChar :: Parser Char
+contentChar =
+    choice
+        [ alphaNumChar
+        , char '.'
+        , char '['
+        , char ']'
+        , char '_'
+        , char '\''
+        , char '"'
+        , char '!'
+        , char '~'
+        , char '-'
+        , char ':'
+        , try escapedParen
+        , char '\\'
+        ]
+
+nonLineSpaceChar :: Parser Char
+nonLineSpaceChar = choice [char ' ', char '\t']
+
+-- This is a replacement for `Text.Megaparsec.Char.Lexer.charLiteral`;
+-- it does nearly the same thing but additionally supports escaped parentheses.
+charLiteral :: Parser Char
+charLiteral = label "literal character" $ do
+    r <- lookAhead (count' 1 10 anySingle)
+    case lexChar r of
+        Just (c, r') -> c <$ skipCount (length r - length r') anySingle
+        Nothing -> unexpected (errorTokens r)
```

Comment:
hm. so if we get `Nothing`, then `lexChar r` must have been called on `r = ""` as `lexChar "" = Nothing` is the only equation which returns `Nothing`. Which means that we are giving `errorTokens ""`, which *also* means that `errorTokens` will call `error "unreachable"` - because `NEL.nonEmpty "" = Nothing`.